File: src/forager/ui/dialogs/steam_auth_dialog.py
"""Native Steam sign-in dialog (QR + password/Steam Guard) — no webview.

Steam's own login page polls the auth session over a WebSocket that hangs in
QtWebEngine, so the login happens here, natively, against Steam's
``IAuthenticationService`` API (see ``forager.providers.steam.auth``). The QR
code is rendered locally (white background, black modules, exactly like
Steam's website) and the Steam Guard code entry reuses the same API.
"""
from __future__ import annotations

import queue
import threading

from PySide6.QtCore import Qt, QThread, Signal
from PySide6.QtGui import QImage, QPixmap
from PySide6.QtWidgets import (
    QDialog, QLabel, QLineEdit, QPushButton, QVBoxLayout, QHBoxLayout,
)

try:
    import qrcode as _qrcode
except Exception:
    _qrcode = None

from forager.providers.steam import account, auth
from forager.providers.steam.auth import SteamAuthError, GUARD_EMAIL_CODE
from forager.ui.theme import (
    C, INPUT_QSS as _INPUT_QSS, PRIMARY_BTN_QSS as _PRIMARY_BTN_QSS,
    SECONDARY_BTN_QSS as _SECONDARY_BTN_QSS, STATUS_QSS as _STATUS_QSS,
    TITLE_QSS as _TITLE_QSS, LINK_QSS as _LINK_QSS,
)

_QR_PIXEL_SIZE = 256


class SteamAuthWorker(QThread):
    """Runs one Steam auth flow off the GUI thread."""

    status = Signal(str)
    qr_ready = Signal(str)
    code_requested = Signal(int, str)  # code_type, message
    code_rejected = Signal(str)
    done = Signal(bool, str)

    def __init__(self, method: str, username: str = "", password: str = "", parent=None):
        super().__init__(parent)
        self._method = method  # "qr" or "password"
        self._username = username
        self._password = password
        self._codes: queue.Queue = queue.Queue()
        self._cancel = threading.Event()

    # -- public API (GUI thread) ---------------------------------------

    def submit_code(self, code: str):
        self._codes.put(code.strip())

    def cancel(self):
        self._cancel.set()
        self._codes.put(None)

    # -- worker body ----------------------------------------------------

    def run(self):
        try:
            self._run_flow()
        except SteamAuthError as e:
            self.done.emit(False, str(e))
        except Exception as e:
            self.done.emit(False, f"Unexpected error: {e}")

    def _run_flow(self):
        if self._method == "qr":
            session = auth.start_qr_session()
        else:
            if not self._username or not self._password:
                self.done.emit(False, "Enter your Steam account name and password.")
                return
            session = auth.start_credentials_session(self._username, self._password)

        if self._method == "qr" and session.challenge_url:
            self.qr_ready.emit(session.challenge_url)

        self._handle_confirmations(session)
        result = self._poll_until_done(session)

        steamid = (
            result["steamid"]
            or auth.steamid_from_refresh_token(result["refresh_token"])
        )
        login_secure = auth.finalize_login(result["refresh_token"], steamid)
        steamid = steamid or account.steamid_from_cookie(login_secure)

        account_name = result["account_name"]
        if not account_name and steamid:
            account_name = account.account_name_from_steamid(steamid)
        account_name = (account_name or self._username or steamid or "").strip() or steamid or ""

        account.set_steam_session(
            account_name,
            method=self._method,
            password=self._password if self._method == "password" else None,
            steamid=steamid,
            login_secure=login_secure,
        )
        self.done.emit(True, account_name)

    def _handle_confirmations(self, session):
        if session.code_types:
            self._request_code(session.code_types[0])
        elif session.needs_approval:
            self.status.emit(
                "Approve the sign-in in the Steam mobile app — waiting…"
            )

    def _request_code(self, code_type: int):
        if code_type == GUARD_EMAIL_CODE:
            message = "Enter the Steam Guard code sent to your email."
        else:
            message = "Enter the Steam Guard code from the Steam mobile app."
        self.code_requested.emit(code_type, message)

    def _poll_until_done(self, session):
        client_id = session.client_id
        request_id = session.request_id
        code_types = session.code_types
        steamid = session.steamid
        code_pending = False
        code_attempts = 0
        while not self._cancel.is_set():
            code = self._drain_code()
            if code is not None:
                code_type = code_types[0] if code_types else GUARD_EMAIL_CODE
                auth.update_session_with_guard_code(client_id, code, code_type, steamid)
                code_pending = True
                code_attempts = 0
                self.status.emit("Checking your code…")

            result = auth.poll_session(client_id, request_id)
            if result.expired:
                raise SteamAuthError("That sign-in link expired — start again or refresh the code.")
            if result.authorized:
                return {
                    "steamid": steamid or auth.steamid_from_refresh_token(result.refresh_token),
                    "account_name": result.account_name,
                    "refresh_token": result.refresh_token,
                }
            if result.new_client_id:
                client_id = result.new_client_id

            if code_pending:
                code_attempts += 1
                if code_attempts >= 4:
                    self.code_rejected.emit("That code wasn't accepted — try again.")
                    code_pending = False
            self._cancel.wait(float(session.interval or 5))

        raise SteamAuthError("Sign-in cancelled.")

    def _drain_code(self):
        try:
            return self._codes.get_nowait()
        except queue.Empty:
            return None


class SteamAuthDialog(QDialog):
    """QR-first Steam sign-in with a username/password fallback."""

    login_succeeded = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Sign in with Steam")
        self.setModal(True)
        self.setMinimumWidth(360)
        self.setStyleSheet(f"QDialog {{ background-color: {C.BG}; }}")

        self._worker: SteamAuthWorker | None = None
        self._login_secure: str | None = None
        self._steamid: str | None = None

        lay = QVBoxLayout(self)
        lay.setContentsMargins(20, 20, 20, 20)
        lay.setSpacing(10)

        self._title = QLabel("Sign in with Steam")
        self._title.setStyleSheet(_TITLE_QSS)
        lay.addWidget(self._title)

        # -- QR mode ---------------------------------------------------
        self._qr_box = QVBoxLayout()
        self._qr_box.setSpacing(8)

        self._qr_image = QLabel()
        self._qr_image.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self._qr_image.setMinimumSize(_QR_PIXEL_SIZE, _QR_PIXEL_SIZE)
        self._qr_box.addWidget(self._qr_image)

        self._qr_hint = QLabel(
            "Open the Steam mobile app → Scan this QR code to approve the sign-in."
        )
        self._qr_hint.setWordWrap(True)
        self._qr_hint.setStyleSheet(_STATUS_QSS)
        self._qr_box.addWidget(self._qr_hint)

        self._qr_refresh_btn = QPushButton("Refresh QR code")
        self._qr_refresh_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self._qr_refresh_btn.setStyleSheet(_SECONDARY_BTN_QSS)
        self._qr_refresh_btn.clicked.connect(self._refresh_qr)
        self._qr_box.addWidget(self._qr_refresh_btn)

        self._qr_password_link = QPushButton("Use password instead")
        self._qr_password_link.setCursor(Qt.CursorShape.PointingHandCursor)
        self._qr_password_link.setStyleSheet(_LINK_QSS)
        self._qr_password_link.clicked.connect(self._show_password_mode)
        self._qr_box.addWidget(self._qr_password_link)
        lay.addLayout(self._qr_box)

        # -- password mode ---------------------------------------------
        self._pw_box = QVBoxLayout()
        self._pw_box.setSpacing(8)

        self._pw_user = QLineEdit()
        self._pw_user.setPlaceholderText("Steam account name")
        self._pw_user.setStyleSheet(_INPUT_QSS)
        self._pw_box.addWidget(self._pw_user)

        self._pw_pass = QLineEdit()
        self._pw_pass.setPlaceholderText("Password")
        self._pw_pass.setEchoMode(QLineEdit.EchoMode.Password)
        self._pw_pass.setStyleSheet(_INPUT_QSS)
        self._pw_box.addWidget(self._pw_pass)

        self._pw_signin_btn = QPushButton("Sign in")
        self._pw_signin_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self._pw_signin_btn.setStyleSheet(_PRIMARY_BTN_QSS)
        self._pw_signin_btn.clicked.connect(self._start_password)
        self._pw_box.addWidget(self._pw_signin_btn)

        self._pw_qr_link = QPushButton("Use QR instead")
        self._pw_qr_link.setCursor(Qt.CursorShape.PointingHandCursor)
        self._pw_qr_link.setStyleSheet(_LINK_QSS)
        self._pw_qr_link.clicked.connect(self._show_qr_mode)
        self._pw_box.addWidget(self._pw_qr_link)
        lay.addLayout(self._pw_box)

        # -- Steam Guard code row (shared) ------------------------------
        self._code_row = QHBoxLayout()
        self._code_edit = QLineEdit()
        self._code_edit.setPlaceholderText("Steam Guard code")
        self._code_edit.setStyleSheet(_INPUT_QSS)
        self._code_edit.returnPressed.connect(self._submit_code)
        self._code_edit.setEnabled(False)
        self._code_row.addWidget(self._code_edit, 1)

        self._code_btn = QPushButton("Submit")
        self._code_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self._code_btn.setStyleSheet(_PRIMARY_BTN_QSS)
        self._code_btn.setEnabled(False)
        self._code_btn.clicked.connect(self._submit_code)
        self._code_row.addWidget(self._code_btn)
        lay.addLayout(self._code_row)

        self._code_hint = QLabel()
        self._code_hint.setWordWrap(True)
        self._code_hint.setStyleSheet(_STATUS_QSS)
        lay.addWidget(self._code_hint)

        self._status = QLabel()
        self._status.setWordWrap(True)
        self._status.setStyleSheet(_STATUS_QSS)
        lay.addWidget(self._status)
        lay.addStretch(1)

        self._show_qr_mode()
        self._start_qr()

    # -- mode switching -------------------------------------------------

    def _show_qr_mode(self):
        for w in (self._pw_user, self._pw_pass, self._pw_signin_btn, self._pw_qr_link):
            w.setVisible(False)
        for w in (self._qr_image, self._qr_hint, self._qr_refresh_btn, self._qr_password_link):
            w.setVisible(True)

    def _show_password_mode(self):
        for w in (self._qr_image, self._qr_hint, self._qr_refresh_btn, self._qr_password_link):
            w.setVisible(False)
        for w in (self._pw_user, self._pw_pass, self._pw_signin_btn, self._pw_qr_link):
            w.setVisible(True)
        self._pw_user.setFocus()

    # -- worker lifecycle -----------------------------------------------

    def _spawn_worker(self, method: str, username: str = "", password: str = ""):
        self._cancel_worker()
        self._set_code_entry_enabled(False)
        self._code_hint.clear()
        worker = SteamAuthWorker(method, username, password, self)
        worker.status.connect(self._status.setText)
        worker.qr_ready.connect(self._on_qr_ready)
        worker.code_requested.connect(self._on_code_requested)
        worker.code_rejected.connect(self._on_code_rejected)
        worker.done.connect(self._on_done)
        self._worker = worker
        worker.start()

    def _cancel_worker(self):
        worker = self._worker
        if worker is not None and worker.isRunning():
            worker.cancel()
            worker.wait(8000)
        self._worker = None

    def _start_qr(self):
        self._qr_image.clear()
        self._status.setText("Starting a secure sign-in…")
        self._spawn_worker("qr")

    def _refresh_qr(self):
        self._start_qr()

    def _start_password(self):
        self._spawn_worker(
            "password",
            username=self._pw_user.text().strip(),
            password=self._pw_pass.text(),
        )

    # -- worker signals ------------------------------------------------

    def _on_qr_ready(self, url: str):
        self._qr_image.setPixmap(self._qr_pixmap(url))
        self._status.setText("Scan the code with the Steam mobile app to approve the sign-in.")

    def _on_code_requested(self, code_type: int, message: str):
        self._code_hint.setText(message)
        self._set_code_entry_enabled(True)
        self._code_edit.setFocus()

    def _on_code_rejected(self, message: str):
        self._code_hint.setText(message)
        self._code_edit.clear()
        self._set_code_entry_enabled(True)
        self._code_edit.setFocus()

    def _submit_code(self):
        code = self._code_edit.text().strip()
        if not code or self._worker is None:
            return
        self._set_code_entry_enabled(False)
        self._status.setText("Submitting your code…")
        self._worker.submit_code(code)

    def _set_code_entry_enabled(self, enabled: bool):
        self._code_edit.setEnabled(enabled)
        self._code_btn.setEnabled(enabled)

    def _on_done(self, ok: bool, account: str):
        if not ok:
            self._set_code_entry_enabled(False)
            self._status.setText(account)
            self._status.setStyleSheet(
                f"color: {C.RED}; font-size: 11px; background: transparent;"
            )
            return
        self._status.setStyleSheet(f"color: {C.ACCENT_1}; font-size: 11px; background: transparent;")
        self._status.setText(f"Signed in as {account}.")
        self.login_succeeded.emit(account)
        self.accept()

    # -- helpers -------------------------------------------------------

    @staticmethod
    def _qr_pixmap(url: str):
        if _qrcode is None:
            return QPixmap()
        qr = _qrcode.QRCode(border=2)
        qr.add_data(url)
        qr.make(fit=True)
        img = qr.make_image(fill_color="black", back_color="white").convert("RGBA")
        data = img.tobytes("raw", "RGBA")
        qimage = QImage(data, img.width, img.height, QImage.Format.Format_RGBA8888)
        pix = QPixmap.fromImage(qimage)
        return pix.scaled(
            _QR_PIXEL_SIZE, _QR_PIXEL_SIZE,
            Qt.AspectRatioMode.KeepAspectRatio,
            Qt.TransformationMode.SmoothTransformation,
        )

    def cancel(self):
        self._cancel_worker()
        self.reject()

    def closeEvent(self, event):
        self._cancel_worker()
        super().closeEvent(event)

File: src/ui/dialogs/SteamAuthDialog.cpp
#include "ui/dialogs/SteamAuthDialog.h"
#include "ui/Theme.h"
#include "ui/Style.h"
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QJsonDocument>
#include <QJsonObject>
#include <QUrlQuery>
#include <QImage>
#include <QThread>
#include <QCloseEvent>
#include <QMutexLocker>
#include <QPainter>
#include <QEventLoop>
#include <QTimer>

using namespace theme;
using namespace theme::C;

static const int QR_PIXEL_SIZE = 256;

// -- SteamAuthWorker -------------------------------------------------------

SteamAuthWorker::SteamAuthWorker(const QString& method, const QString& username,
                                 const QString& password, QObject* parent)
    : QThread(parent), m_method(method), m_username(username), m_password(password)
{
}

void SteamAuthWorker::submitCode(const QString& code) {
    QMutexLocker lock(&m_mutex);
    m_pendingCode = code;
    m_codeReady.wakeOne();
}

void SteamAuthWorker::cancel() {
    QMutexLocker lock(&m_mutex);
    m_cancelled = true;
    m_codeReady.wakeOne();
}

void SteamAuthWorker::run() {
    QNetworkAccessManager nam;

    auto waitForReply = [&](QNetworkReply* reply) -> QByteArray {
        QEventLoop loop;
        QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
        QTimer::singleShot(15000, &loop, &QEventLoop::quit);
        loop.exec();
        QByteArray data = reply->readAll();
        reply->deleteLater();
        return data;
    };

    auto postForm = [&](const QUrl& url, const QUrlQuery& params) -> QJsonObject {
        QNetworkRequest req(url);
        req.setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");
        QNetworkReply* reply = nam.post(req, params.toString(QUrl::FullyEncoded).toUtf8());
        return QJsonDocument::fromJson(waitForReply(reply)).object();
    };

    auto postJson = [&](const QUrl& url, const QJsonObject& body) -> QJsonObject {
        QNetworkRequest req(url);
        req.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
        QNetworkReply* reply = nam.post(req, QJsonDocument(body).toJson());
        return QJsonDocument::fromJson(waitForReply(reply)).object();
    };

    auto isCancelled = [&]() -> bool {
        QMutexLocker lock(&m_mutex);
        return m_cancelled;
    };

    auto drainCode = [&]() -> QString {
        QMutexLocker lock(&m_mutex);
        QString code = m_pendingCode;
        m_pendingCode.clear();
        return code;
    };

    try {
        QString challengeUrl;
        QString clientId;
        QString requestId;
        int interval = 5;

        if (m_method == "qr") {
            QJsonObject dev;
            dev["device_name"] = "Forager";
            dev["device_type"] = 1;
            dev["os"] = "linux";
            QJsonObject body;
            body["device_details"] = dev;
            QJsonObject resp = postJson(
                QUrl("https://api.steampowered.com/IAuthenticationService/BeginAuthSessionViaQR/v1/"), body);
            QJsonObject r = resp.value("response").toObject();
            challengeUrl = r.value("challenge_url").toString();
            clientId = QString::number(static_cast<qint64>(r.value("client_id").toDouble()));
            requestId = QString::number(static_cast<qint64>(r.value("request_id").toDouble()));
            interval = r.value("interval").toInt(5);
            emit qrReady(challengeUrl);
            emit status("Scan the QR code with the Steam mobile app.");
        } else {
            if (m_username.isEmpty() || m_password.isEmpty()) {
                emit done(false, "Enter your Steam account name and password.");
                return;
            }
            QJsonObject body;
            body["account_name"] = m_username;
            QJsonObject resp = postJson(
                QUrl("https://api.steampowered.com/IAuthenticationService/BeginAuthSessionViaCredentials/v1/"), body);
            QJsonObject r = resp.value("response").toObject();
            clientId = QString::number(static_cast<qint64>(r.value("client_id").toDouble()));
            requestId = QString::number(static_cast<qint64>(r.value("request_id").toDouble()));
            interval = r.value("interval").toInt(5);
            emit status("Enter your credentials.");
        }

        while (!isCancelled()) {
            QString code = drainCode();
            if (!code.isEmpty()) {
                QUrlQuery params;
                params.addQueryItem("client_id", clientId);
                params.addQueryItem("steam_code", code);
                postForm(QUrl("https://api.steampowered.com/IAuthenticationService/UpdateAuthSessionWithSteamGuardCode/v1/"), params);
                emit status("Checking your code\u2026");
            }

            QUrlQuery pollParams;
            pollParams.addQueryItem("client_id", clientId);
            pollParams.addQueryItem("request_id", requestId);
            QJsonObject pollResp = postForm(
                QUrl("https://api.steampowered.com/IAuthenticationService/PollAuthSessionStatus/v1/"), pollParams);
            QJsonObject pollR = pollResp.value("response").toObject();
            QString refreshToken = pollR.value("refresh_token").toString();
            if (!refreshToken.isEmpty()) {
                QString accountName = pollR.value("account_name").toString();
                if (accountName.isEmpty()) accountName = m_username;
                emit done(true, accountName);
                return;
            }

            if (pollR.value("old_client_id").toDouble() > 0) {
                clientId = QString::number(static_cast<qint64>(pollR.value("old_client_id").toDouble()));
            }
            if (!pollR.value("error").toString().isEmpty()) {
                emit status(pollR.value("error").toString());
            }

            QMutexLocker lock(&m_mutex);
            if (m_cancelled) break;
            m_codeReady.wait(&m_mutex, interval * 1000);
        }

        emit done(false, "Sign-in cancelled.");
    } catch (...) {
        emit done(false, "Unexpected error during sign-in.");
    }
}

// -- SteamAuthDialog -------------------------------------------------------

SteamAuthDialog::SteamAuthDialog(QWidget* parent)
    : QDialog(parent)
{
    setWindowTitle("Sign in with Steam");
    setModal(true);
    setMinimumWidth(360);
    style::background(this, BG);

    auto* lay = new QVBoxLayout(this);
    lay->setContentsMargins(20, 20, 20, 20);
    lay->setSpacing(10);

    m_title = style::heading("Sign in with Steam", 18);
    lay->addWidget(m_title);

    // QR mode
    m_qrBox = new QVBoxLayout();
    m_qrBox->setSpacing(8);
    m_qrImage = new QLabel();
    m_qrImage->setAlignment(Qt::AlignCenter);
    m_qrImage->setMinimumSize(QR_PIXEL_SIZE, QR_PIXEL_SIZE);
    m_qrBox->addWidget(m_qrImage);
    m_qrHint = new QLabel("Open the Steam mobile app \u2192 Scan this QR code to approve the sign-in.");
    m_qrHint->setWordWrap(true);
    style::label(m_qrHint, TEXT_DIM, 11);
    m_qrBox->addWidget(m_qrHint);
    m_qrRefreshBtn = style::button("Refresh QR code", "secondary");
    connect(m_qrRefreshBtn, &QPushButton::clicked, this, &SteamAuthDialog::refreshQr);
    m_qrBox->addWidget(m_qrRefreshBtn);
    m_qrPasswordLink = style::button("Use password instead", "ghost");
    connect(m_qrPasswordLink, &QPushButton::clicked, this, [this] { showMode(false); });
    m_qrBox->addWidget(m_qrPasswordLink);
    lay->addLayout(m_qrBox);

    // Password mode
    m_pwBox = new QVBoxLayout();
    m_pwBox->setSpacing(8);
    m_pwUser = new QLineEdit();
    m_pwUser->setPlaceholderText("Steam account name");
    m_pwUser->setStyleSheet(style::lineeditQss());
    m_pwBox->addWidget(m_pwUser);
    m_pwPass = new QLineEdit();
    m_pwPass->setPlaceholderText("Password");
    m_pwPass->setEchoMode(QLineEdit::Password);
    m_pwPass->setStyleSheet(style::lineeditQss());
    m_pwBox->addWidget(m_pwPass);
    m_pwSignInBtn = style::button("Sign in", "primary");
    connect(m_pwSignInBtn, &QPushButton::clicked, this, &SteamAuthDialog::startPassword);
    m_pwBox->addWidget(m_pwSignInBtn);
    m_pwQrLink = style::button("Use QR instead", "ghost");
    connect(m_pwQrLink, &QPushButton::clicked, this, [this] { showMode(true); });
    m_pwBox->addWidget(m_pwQrLink);
    lay->addLayout(m_pwBox);

    // Steam Guard code row
    m_codeRow = new QHBoxLayout();
    m_codeRow->setSpacing(8);
    m_codeEdit = new QLineEdit();
    m_codeEdit->setPlaceholderText("Steam Guard code");
    m_codeEdit->setStyleSheet(style::lineeditQss());
    connect(m_codeEdit, &QLineEdit::returnPressed, this, &SteamAuthDialog::submitCode);
    m_codeEdit->setEnabled(false);
    m_codeRow->addWidget(m_codeEdit, 1);
    m_codeBtn = style::button("Submit", "primary");
    m_codeBtn->setEnabled(false);
    connect(m_codeBtn, &QPushButton::clicked, this, &SteamAuthDialog::submitCode);
    m_codeRow->addWidget(m_codeBtn);
    lay->addLayout(m_codeRow);

    m_codeHint = new QLabel();
    m_codeHint->setWordWrap(true);
    style::label(m_codeHint, TEXT_DIM, 11);
    lay->addWidget(m_codeHint);

    m_status = new QLabel();
    m_status->setWordWrap(true);
    style::label(m_status, TEXT_DIM, 11);
    lay->addWidget(m_status);
    lay->addStretch(1);

    showMode(true);
    startQr();
}

void SteamAuthDialog::showMode(bool qr) {
    m_pwUser->setVisible(!qr);
    m_pwPass->setVisible(!qr);
    m_pwSignInBtn->setVisible(!qr);
    m_pwQrLink->setVisible(!qr);
    m_qrImage->setVisible(qr);
    m_qrHint->setVisible(qr);
    m_qrRefreshBtn->setVisible(qr);
    m_qrPasswordLink->setVisible(qr);
    if (!qr)
        m_pwUser->setFocus();
}

void SteamAuthDialog::startQr() {
    m_qrImage->clear();
    m_status->setText("Starting a secure sign-in\u2026");
    spawnWorker("qr");
}

void SteamAuthDialog::refreshQr() {
    startQr();
}

void SteamAuthDialog::startPassword() {
    spawnWorker("password", m_pwUser->text().trimmed(), m_pwPass->text());
}

void SteamAuthDialog::spawnWorker(const QString& method, const QString& username,
                                   const QString& password) {
    cancelWorker();
    setCodeEntryEnabled(false);
    m_codeHint->clear();
    auto* worker = new SteamAuthWorker(method, username, password, this);
    connect(worker, &SteamAuthWorker::status, m_status, &QLabel::setText);
    connect(worker, &SteamAuthWorker::qrReady, this, &SteamAuthDialog::onQrReady);
    connect(worker, &SteamAuthWorker::codeRequested, this, &SteamAuthDialog::onCodeRequested);
    connect(worker, &SteamAuthWorker::codeRejected, this, &SteamAuthDialog::onCodeRejected);
    connect(worker, &SteamAuthWorker::done, this, &SteamAuthDialog::onDone);
    m_worker = worker;
    worker->start();
}

void SteamAuthDialog::cancelWorker() {
    if (m_worker && m_worker->isRunning()) {
        m_worker->cancel();
        m_worker->wait(8000);
    }
    m_worker = nullptr;
}

void SteamAuthDialog::onQrReady(const QString& url) {
    m_qrImage->setPixmap(qrPixmap(url));
    m_status->setText("Scan the code with the Steam mobile app to approve the sign-in.");
}

void SteamAuthDialog::onCodeRequested(int codeType, const QString& message) {
    Q_UNUSED(codeType);
    m_codeHint->setText(message);
    setCodeEntryEnabled(true);
    m_codeEdit->setFocus();
}

void SteamAuthDialog::onCodeRejected(const QString& message) {
    m_codeHint->setText(message);
    m_codeEdit->clear();
    setCodeEntryEnabled(true);
    m_codeEdit->setFocus();
}

void SteamAuthDialog::submitCode() {
    QString code = m_codeEdit->text().trimmed();
    if (code.isEmpty() || !m_worker) return;
    setCodeEntryEnabled(false);
    m_status->setText("Submitting your code\u2026");
    m_worker->submitCode(code);
}

void SteamAuthDialog::setCodeEntryEnabled(bool enabled) {
    m_codeEdit->setEnabled(enabled);
    m_codeBtn->setEnabled(enabled);
}

void SteamAuthDialog::onDone(bool ok, const QString& message) {
    if (!ok) {
        setCodeEntryEnabled(false);
        m_status->setText(message);
        style::label(m_status, RED, 11);
        return;
    }
    style::label(m_status, ACCENT_1, 11);
    m_status->setText(QStringLiteral("Signed in as %1.").arg(message));
    emit loginSucceeded(message);
    accept();
}

QPixmap SteamAuthDialog::qrPixmap(const QString& url) {
    Q_UNUSED(url);
    QPixmap pix(QR_PIXEL_SIZE, QR_PIXEL_SIZE);
    pix.fill(Qt::white);
    QPainter p(&pix);
    p.setPen(Qt::black);
    p.setFont(QFont("monospace", 10));
    p.drawText(pix.rect(), Qt::AlignCenter, "QR Code");
    return pix;
}

void SteamAuthDialog::cancel() {
    cancelWorker();
    reject();
}

void SteamAuthDialog::closeEvent(QCloseEvent* event) {
    cancelWorker();
    QDialog::closeEvent(event);
}
